pkg/feeds/sequence/sequence.go

```diff
@@ -0,0 +1,208 @@
+package sequence
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/feeds"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+type index struct {
+	index uint64
+}
+
+func (i *index) MarshalBinary() ([]byte, error) {
+	indexBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(indexBytes, i.index)
+	return indexBytes, nil
+}
+
+// Finder encapsulates a chunk store getter and a feed and provides
+//  non-concurrent lookup methods
+type Finder struct {
+	getter *feeds.Getter
+}
+
+// NewFinder constructs an Finder
+func NewFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
+	return &Finder{feeds.NewGetter(getter, feed)}
+}
+
+// At looks up the version valid at time `at`
+// after is a unix time hint of the latest known update
+func (f *Finder) At(ctx context.Context, at, after int64) (ch swarm.Chunk, err error) {
+	for i := uint64(1); ; i++ {
+		u, err := f.getter.Get(ctx, &index{i})
+		if err != nil {
+			if !errors.Is(err, storage.ErrNotFound) {
+				return nil, err
+			}
+			return ch, nil
+		}
+		ts, err := feeds.UpdatedAt(u)
+		if err != nil {
+			return nil, err
+		}
+		if ts > uint64(at) {
+			return ch, nil
+		}
+		ch = u
+	}
+}
+
+// Finder encapsulates a chunk store getter and a feed and provides
+//  non-concurrent lookup methods
+type AsyncFinder struct {
+	getter *feeds.Getter
+}
+
+// NewAsyncFinder constructs an AsyncFinder
+func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
+	return &AsyncFinder{feeds.NewGetter(getter, feed)}
+}
+
+type path struct {
+	latest result
+	base   uint64
+	level  int
+	cancel chan struct{}
+}
+
+func newPath(base uint64) *path {
+	return &path{base: base, cancel: make(chan struct{})}
+}
+
+type result struct {
+	chunk swarm.Chunk
+	path  *path
+	level int
+	seq   uint64
+	diff  int64
+}
+
+// At looks up the version valid at time `at`
+// after is a unix time hint of the latest known update
+func (f *AsyncFinder) At(ctx context.Context, at, after int64) (ch swarm.Chunk, err error) {
+	ch, diff, err := f.get(ctx, at, 1)
```

Comment:
W

Reply:
it was convenient but it is unexpected and inconsistent, so i changed  it to 0 start